### PR TITLE
Improved data re-use instructions

### PR DESCRIPTION
**Pull request by duncan-brown**
This is a re-write of the data re-use instructions to explain how re-use works and give a few common examples.

A rendered version of the changes can be found [here.](https://github.com/duncan-brown/pycbc/blob/reuse_doc/docs/workflow/pycbc_make_coinc_search_workflow.rst#reuse-of-data-from-a-previous-workflow)

### CODE REVIEW

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 1]
```diff
@@ -608,53 +608,238 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
-Select the entries in the file that you would like to skip generating again and
-place that into a new file. The example below selects all the inspiral and 
-tmpltbank jobs and places their entries into a new listing called prior_data.map.::
-
-    # Lets get the tmpltbank entries
-    cat /path/to/old/run/${GPS_START_TIME}-${GPS_END_TIME}/output.map | grep 'TMPLTBANK' > prior_data.map
-    
-    # Add in the inspiral  files
-    cat /path/to/old/run/${GPS_START_TIME}-${GPS_END_TIME}/output.map | grep 'INSPIRAL' >> prior_data.map
+One of the features of Pegasus is reus the data products of prior runs.
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: reus -> reuse

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 2]
```diff
@@ -608,53 +608,238 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+This can be used to e.g. expand an analysis or recover a run with mistaken settings without
+duplicating work. The steps below explain how to do this.
+
+------------------------------------
+Setting up a workflow for data reuse
+------------------------------------
+
+The first step is to generate a new workflow that performs the analysis that
+you would like to do. Data reuse happens at the ``pycbc_submit_dax`` step, so
+first run ``pycbc_make_coinc_search_workflow`` to build a new workflow,
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Suggest to say explicitly that this should be in a new directory

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 3]
```diff
@@ -608,53 +608,238 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+following the instructions in the section :ref:`coincworkflowgenerate` of this
+page.
+
+**Stop** before you plan and submit the workflow with ``pycbc_submit_dax``.
+You will pass an additional file to ``pycbc_submit_dax`` using the
+``--cache-file`` option with a list of files that Pegasus can re-use from a
+pervious run.  The Pegasus Workflow Planner will try to reduce the workflow
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: pervious -> previous

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 1 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 2 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 3 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Fixed

**Comment by tdent** [comment]
Hi Duncan, I hadn't got more than 20% through looking at this (you posted it late yesterday evening and this morning was the weekly AEI CBC group meeting).  Do you want me to review it properly?

**Comment by duncan-brown** [comment]
@tdent yes, please go through it carefully. I would rather wait to merge until someone has looked in detail. I just expanded the veto re-use section, but won't make any more updates until you review.

It would be good if @ahnitz and @spxiwh can also take a look.

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 4]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+previous run.  The Pegasus Workflow Planner will reduce the workflow
+using this cache file. Reduction works by deleting jobs from the workflow
+whose output files have been found in some location in this cache file.
+
+The key to data reuse is building the cache file passed to ``pycbc_submit_dax``. This file maps a file created in the workflow to a URL and a site where that URL can be found. The syntax of the cache file is plain ASCII with each line in the file giving the location of a file in the format::
+
+    LOGICAL_FILE_NAME PHYSICAL_FILE_URL pool="SITE"
+
+where ``LOGICAL_FILE_NAME`` is the name of the file as it appears in the
+workflow (incluing any preceeding path), ``PHYSICAL_FILE_URL`` is a URL where
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: typo 'incluing' and preceeding -> preceding

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 5]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+the file can be found, and ``SITE`` is the site on which that URL resides.
+
+The URI in the ``PHYSICAL_FILE_URL`` can be any of the URIs that Pegasus
+recognizes. The URIs ``file://``, ``gsiftp://``, and ``http://`` are likely
+the most useful. Pegasus will take care of adding transfer jobs for
+``gsiftp://`` and ``http://`` URIs, if the data is not available locally.
+
+The string ``SITE`` is a hint that tells Pegasus on which site the
+``PHYSICAL_FILE_URL`` can be found (This is needed beacuse the execution site
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: i don't know what 'execution site' means here. either de-jargonize or omit the explanation

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 6]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+    L1-VETOTIME_CAT3-1169107218-1066800.xml file://localhost/home/dbrown/projects/aligo/o2/analysis-4/o2-analysis-4/output/results/1._analysis_time/1.01_segment_data/L1-VETOTIME_CAT3-1169107218-1066800.xml pool="local"
+    116912/H1-INSPIRAL_FULL_DATA_JOB0-1169120586-1662.hdf file://localhost/home/dbrown/projects/aligo/o2/analysis-4/o2-analysis-4/output/full_data/H1-INSPIRAL_FULL_DATA_JOB0-1169120586-1662.hdf pool="local"
+    116912/H1-INSPIRAL_FULL_DATA_JOB1-1169120586-1662.hdf file://localhost/home/dbrown/projects/aligo/o2/analysis-4/o2-analysis-4/output/full_data/H1-INSPIRAL_FULL_DATA_JOB1-1169120586-1662.hdf pool="local"
+
+In this case, Pegasus will delete the jobs from the workflow that create the files ``H1-VETOTIME_CAT3-1169107218-1066800.xml``, ``L1-VETOTIME_CAT3-1169107218-1066800.xml``, ``116912/H1-INSPIRAL_FULL_DATA_JOB0-1169120586-1662.hdf``, and ``116912/H1-INSPIRAL_FULL_DATA_JOB1-1169120586-1662.hdf`` when it plans the workflow. Insted, the data will be re-used from the URLs specified in the cache. Since ``site="local"`` for these files, Pegasus expects that the files all exist on the host where the workflow is run from.
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: typo 'insted'

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 7]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+than ``pool="local"``. Pegasus will re-use these data files adding in
+file transfer jobs to the workflow to move them into the appropriate
+locations.
+
+Once a cache file has been constructed, to enable data re-use, you follow the
+standard instructions for planning and submitting the workflow in the section
+:ref:`coincworkflowplan`, but add the ``--cache-file`` argument that points to
+the cache file that you have created. For example:: 
+
+    pycbc_submit_dax --accounting-group ligo.dev.o1.cbc.explore.test --dax s6d_chunk3.dax --cache-fole /path/to/prior_data.map
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: typo 'fole' -> file (!)

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 8]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+
+will use the URLs from the file ``/path/to/prior_data.map`` to implement
+data re-use and subsequent workflow reduction.
+
+The main key to data re-use is generating the cache file. There is no rule for
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: suggest to rephrase : 'Which file URLs should be included in the reuse cache?  There is no single correct way of deciding this ...'

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 9]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+doing this, as it depends on eactly what you are trying to do. The sections
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: typo 'eactly'

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 10]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+below explain how to do this for a few common situations.
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: rephrase 'how to select the appropriate files to reuse ..'

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 11]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
 
 .. note::
 
-    You can include files in the prior data listing that wouldn't be generated anyway by your new run. These are simply
-    ignored.
+    The ``[workflow]`` section of the ini configuration file contains an
+    option ``file-retention-level``. This is commonly set to ``all_files`` or
+    ``all_triggers``, in which was the data products re-used will be copied
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: in which was -> in which case?

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 12]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
 
-From the directory where the dax was created, now plan the workflow with an additional argument as follows.::
+    sed 's/pool="local"/pool="remote"/g' inspiral_files.map > inspiral_files.map.tmp
+    sed 's+/home/dbrown+gsiftp://sugwg-condor.phy.syr.edu/home/dbrown+g' inspiral_files.map.tmp > inspiral_files.map
+    rm inspiral_files.map.tmp
 
-    pycbc_submit_dax --accounting-group ligo.dev.o1.cbc.explore.test --dax s6d_chunk3.dax --cache /path/to/prior_data.map
+6. Finally, copy the file ``inspiral_files.map`` to your new workflow directory and then run ``pycbc_submit_dax`` as usual, giving the path to ``inspiral_files.map`` as the ``--cache-file`` argument.
+
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: maybe note that in this case you do *not* change the retention level, it has to stay at 'all_triggers'

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 13]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+---------------------------------------------------
+Re-running a workflow using a new veto definer file
+---------------------------------------------------
+
+Data reuse can be used to re-running a workflow with a new veto definer file, assuming that:
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: it's not clear to me that the calculate_psd and optimal_snr jobs will always be valid after the veto definer is changed - in principle if there are new significant CAT1 vetoes the psds could have substantial changes.  if there are only CAT2+ veto changes then they are OK (and so are the hdf_trigger_merge files ..)

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 14]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+In this case, first re-run ``pycbc_make_coinc_search_workflow`` to build the
+new workflow, and then create a cache file in the following way:
+
+1. Locate the PyCBC result page for the workflow that you wish to extend.
+
+2. In the menu under **Section 8: Workflow**, locate the **Output map** section (usually Section 8.06) and open that page.
+
+3. This page will show three output cache files that contain the URLs of the data created by the workflow. Locate the file that ends ``main.map`` and download it by clicking on the **Link to file**. This file contains the main intermediate and output data products of the workflow.
+
+4. Remove output files that match the following strings from the output map file:
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: what about the hdf_trigger_merge files

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 15]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+
+    egrep -v '(VETOTIME|LIGOLW_COMBINE_SEGMENTS|CUMULATIVE_CAT_12H_VETO_SEGMENTS|COINC|FIT|STATMAP|INJFIND|PAGE|FOREGROUND_CENSOR|html|png|dax)' /path/to/main.map > /path/to/reuse_cache.map
+
+6. Copy the file ``reuse_cache.map`` to your new workflow directory and then run ``pycbc_submit_dax`` as usual, giving the path to ``reuse_cache.map`` as the ``--cache-file`` argument.
+
+----------------------------
+Re-running a failed workflow
+----------------------------
+
+Ocasionaly it may be necessary to use data from a partially completed
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Occasionally (sp.)

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 16]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+the ``osg-site-scratch`` directory. In this case::
+
+    cd /path/to/workflow/output/osg-site-scratch/work
+
+and change into the directory that ends with ``main_ID0000001``. Then run the
+same command as above::
+
+    for pfn in `find . -type f | sed 's+^./++g'` ; do echo $pfn file://`pwd`/$pfn pool=\"local\" ; done >> /path/to/partial_workflow.map
+
+but **note** the ``>>`` rather than ``>`` to append to the file ``partial_workflow.map``, rather than clobbering the file and creating a new file.
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: clobbering -> deleting the existing file?

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 17]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+
+------------------------------------
+Setting up a workflow for data reuse
+------------------------------------
+
+The first step is to generate a new workflow that performs the analysis that
+you would like to do. This workflow should be generated in a new directory so that it does not overwrite data from your previous workflows.
+Data reuse happens at the ``pycbc_submit_dax`` step, so
+first run ``pycbc_make_coinc_search_workflow`` to build a new workflow,
+following the instructions in the section :ref:`coincworkflowgenerate` of this
```
Comment (spxiwh) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: The ref here didn't work in the compiled documentation ... Why didn't that cause a failure in the docs build??

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 18]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+If you are changing the configuration parameters of a workflow, then you can
+build a cache file from a previous ``main.map`` file or the files under
+``local-site-scratch``, but you will need to filter the cache file to remove
+the files for jobs that have a changed configuration.  Here are a few
+examples:
+
+* If you are changing the configuration of ``pycbc_inspiral`` you must regenerate almost all the files in the workflow so it easier to start from scratch.
+* If you are changing the injections, but want to re-use the ``FULL_DATA`` previous analysis, you can filter the ``main.map`` to keep the veto files, template bank files, full data inspiral files, and PSD files. For example::
+
+    egrep '(VETO|BANK|FULL_DATA|PSD)' /path/to/main.map > /path/to/reuse.map
```
Comment (tdent) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: will this pick up some full data '.dax' files?  I have not found that reuse works properly if .dax files are in the cache

**Comment by tdent** [comment]
ok, i've gone through the text, there are a few substantive questions on the data reuse

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 19]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+
+Note that the URL now contains ``gsiftp://sugwg-condor.phy.syr.edu`` rather
+than ``file://localhost`` and the files are listes as ``pool="remote"`` rather
+than ``pool="local"``. Pegasus will re-use these data files adding in
+file transfer jobs to the workflow to move them into the appropriate
+locations.
+
+Once a cache file has been constructed, to enable data re-use, you follow the
+standard instructions for planning and submitting the workflow in the section
+:ref:`coincworkflowplan`, but add the ``--cache-file`` argument that points to
```
Comment (spxiwh) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Here too.

File: docs/workflow/pycbc_make_coinc_search_workflow.rst [hunk 20]
```diff
@@ -608,53 +608,250 @@ If you have a subdax that failed, ``pegasus_analyzer`` will provide you with a c
...
+    The ``[workflow]`` section of the ini configuration file contains an
+    option ``file-retention-level``. This is commonly set to ``all_files`` or
+    ``all_triggers``, in which was the data products re-used will be copied
+    from the input locations and stored into the output location of the new
+    workflow when the new workflow is run with data re-use. This can be
+    wasteful of disk space, so you may want to set this option to either
+    ``merged_triggers`` or ``results`` to store a smaller sub-set of the
+    workflow's data products. These setting will allow the use of data from
+    a previous run, but not make duplicate copies of intermediate data files.
+    See the documentation under :ref:`workflowconfigparsermod` for more
```
Comment (spxiwh) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: And here ... (the generator really should have failed).

**Comment by spxiwh** [comment]
This is branched quite a way from master. Can you rebase this to demonstrate that the docs build without failure. (ie. A requirement for docs changes is to actually build the docs, which you've done, but as the code version seems so old, I'm not sure it's reporting issues for e.g. the broken links)

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 4 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 5 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Added additional explanation of what ``SITE`` means.

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 6 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 7 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Fixed.

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 8 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 10 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 11 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 12 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Added discussion of ``file-retention-level`` here.

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 13 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Added:

```
If category 1 vetoes have changed, you must also remove files matching ``PSD``, ``OPTIMAL``, and ``MERGE`` to remove the PSD estimation jobs, the jobs that compute the optimal SNR of injections, and the merged single-detector inspiral trigger files which may also change if the category 1 vetoes change.
```

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 14 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: they need to be removed if CAT1 changes. added note

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 15 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 16 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: fixed

File: docs/workflow/pycbc_make_coinc_search_workflow.rst (hunk 18 above)
Comment (duncan-brown) [inline docs/workflow/pycbc_make_coinc_search_workflow.rst]: Good point,fixed.

**Comment by duncan-brown** [comment]
@tdent comments address, thanks. Please approve if you are happy and I will merge.

@spxiwh the links only work once it is rendered in Sphinx.

**Comment by spxiwh** [comment]
@duncan-brown Oh yeah, I hadn't realised this was a link to the rst file, not the Sphinx rendered stuff. Can you verify that this does build correctly in Sphinx?

**Comment by duncan-brown** [comment]
@spxiwh yep

http://duncan-brown.github.io/pycbc/latest/html/workflow/pycbc_make_coinc_search_workflow.html#reuse-of-data-from-a-previous-workflow

**Comment by duncan-brown** [comment]
I think Codacy is wedged, so I merged and rebuilt the docs:

http://ligo-cbc.github.io/pycbc/latest/html/workflow/pycbc_make_coinc_search_workflow.html#reuse-of-data-from-a-previous-workflow

**Pull request merged**